aptos-move/framework/aptos-framework/sources/aggregator/aggregator.spec.move

```diff
@@ -6,20 +6,24 @@ spec aptos_framework::aggregator {
     spec add {
         // TODO: temporary mockup.
         pragma opaque;
+        aborts_if false;
     }
 
     spec sub {
         // TODO: temporary mockup.
         pragma opaque;
+        aborts_if false;
     }
 
     spec read {
         // TODO: temporary mockup.
         pragma opaque;
+        aborts_if false;
```

Comment:
@georgemitenkov , can there be any cases where `read` and `destroy` may potentially abort?

Reply:
`destroy` never aborts, and so should `read`. Say you have an aggregator with an overflow limit of 10. And some transaction does this:

```
a = Aggregator(limit=10) // actual storage value is 5 already, but this transaction doesn't know
a.add(7)
a.read() // aborts the transaction since we read 5+7 > 10
```
So technically `read` aborts because it checks if previous deltas made sense. At the same time,
```
a = Aggregator(limit=10) // actual storage value is 5 already, but this transaction doesn't know
a.add(7)
```
equally aborts because at the end of the transaction we check if it should have been aborted (e.g. in sequential execution this means read the value, in parallel execution it is a bit more convoluted).

So, in my opinion, this shows that `read` never aborts, but rather catches an abort in `add`/`sub`. Not sure how you would qualify this for the prover?

Reply:
Looking at native implementation, we can abort if aggregator value is not found in storage, do we count this is as well? Technically should never happen...

Reply:
@qpb8023 , I am confused too. Does Prover still produce this error? Or, is this error gone after you add `aborts_if false` to the spec of `aggregator::destroy`?

Reply:
The `aggregator::destroy` function is called by `optional_aggregator::switch_to_integer_and_zero_out` and `optional_aggregator::destroy_optional_aggregator`.


with below spec , `optional_aggregator::switch_to_integer_and_zero_out` and `optional_aggregator::destroy_optional_aggregator` don't abort .
```
    spec destroy {
        // TODO: temporary mockup.
        pragma opaque;
        aborts_if false;
    }
```

with below spec , `optional_aggregator::switch_to_integer_and_zero_out` and `optional_aggregator::destroy_optional_aggregator` aborts .
```
    spec destroy {
        // TODO: temporary mockup.
        pragma opaque;
        //aborts_if false;
    }
```

**the steps to reproduce is below:**

1  execute 1 of the below commands
```
aptos move prove --package-dir /root/movebit/aptos/movebit/aptos-core/aptos-move/framework/aptos-framework -f  aggregator.move

./move prove -p /root/movebit/aptos/movebit/aptos-core/aptos-move/framework/aptos-framework -t aggregator. Move
```

2 got error results, it seems the `aggregator::destroy` will abort in some cases.

```
[INFO] preparing module 0x1::aggregator
[INFO] preparing module 0x1::optional_aggregator
[INFO] transforming bytecode
[INFO] generating verification conditions
[INFO] 15 verification conditions
[INFO] running solver
[INFO] 0.318s build, 0.257s trafo, 0.007s gen, 1.166s verify, total 1.750s
error: abort not covered by any of the `aborts_if` clauses
    ┌─ ./sources/aggregator/optional_aggregator.spec.move:67:5
    │
 67 │ ╭     spec destroy(optional_aggregator: OptionalAggregator) {
 68 │ │         aborts_if is_parallelizable(optional_aggregator) && len(optional_aggregator.integer.vec) != 0;
 69 │ │         aborts_if !is_parallelizable(optional_aggregator) && len(optional_aggregator.integer.vec) == 0;
 70 │ │     }
    │ ╰─────^
    │
    ┌─ ./sources/aggregator/optional_aggregator.move:137:9
    │
137 │         aggregator::destroy(option::destroy_some(aggregator));
    │         ----------------------------------------------------- abort happened here with code 0x2
    │
    =     at ./sources/aggregator/optional_aggregator.move:125: destroy
    =         optional_aggregator =
    =           optional_aggregator.OptionalAggregator{
    =             aggregator =
    =               option.Option{
    =                 vec =
    =                   vector{
    =                     aggregator.Aggregator{
    =                       handle = 0x18be,
    =                       key = 0x4823,
    =                       limit = 41u128}}},
    =             integer = option.Option{vec = vector{}}}
    =     at ./sources/aggregator/optional_aggregator.move:126: destroy
    =     at ./sources/aggregator/optional_aggregator.move:185: is_parallelizable
    =         optional_aggregator =
    =           optional_aggregator.OptionalAggregator{
    =             aggregator =
    =               option.Option{
    =                 vec =
    =                   vector{
    =                     aggregator.Aggregator{
    =                       handle = 0x18be,
    =                       key = 0x4823,
    =                       limit = 41u128}}},
    =             integer = option.Option{vec = vector{}}}
    =     at ./sources/aggregator/optional_aggregator.move:186: is_parallelizable
    =         result = true
    =     at ./sources/aggregator/optional_aggregator.move:187: is_parallelizable
    =     at ./sources/aggregator/optional_aggregator.move:127: destroy
    =     at ./sources/aggregator/optional_aggregator.move:134: destroy_optional_aggregator
    =         optional_aggregator =
    =           optional_aggregator.OptionalAggregator{
    =             aggregator =
    =               option.Option{
    =                 vec =
    =                   vector{
    =                     aggregator.Aggregator{
    =                       handle = 0x18be,
    =                       key = 0x4823,
    =                       limit = 41u128}}},
    =             integer = option.Option{vec = vector{}}}
    =     at ./sources/aggregator/optional_aggregator.move:135: destroy_optional_aggregator
    =         integer = option.Option{vec = vector{}}
    =         aggregator =
    =           option.Option{
    =             vec =
    =               vector{
    =                 aggregator.Aggregator{
    =                   handle = 0x18be,
    =                   key = 0x4823,
    =                   limit = 41u128}}}
    =     at ./sources/aggregator/optional_aggregator.move:136: destroy_optional_aggregator
    =     at ./sources/aggregator/aggregator.move:80: limit
    =         aggregator =
    =           aggregator.Aggregator{handle = 0x18be, key = 0x4823, limit = 41u128}
    =     at ./sources/aggregator/aggregator.move:81: limit
    =         result = 41u128
    =     at ./sources/aggregator/aggregator.move:82: limit
    =         limit = 41u128
    =     at ./sources/aggregator/optional_aggregator.move:137: destroy_optional_aggregator
    =         ABORTED

error: abort not covered by any of the `aborts_if` clauses
    ┌─ ./sources/aggregator/optional_aggregator.spec.move:73:5
    │
 73 │ ╭     spec destroy_optional_aggregator(optional_aggregator: OptionalAggregator): u128 {
 74 │ │         aborts_if len(optional_aggregator.aggregator.vec) == 0;
 75 │ │         aborts_if len(optional_aggregator.integer.vec) != 0;
 76 │ │     }
    │ ╰─────^
    │
    ┌─ ./sources/aggregator/optional_aggregator.move:137:9
    │
137 │         aggregator::destroy(option::destroy_some(aggregator));
    │         ----------------------------------------------------- abort happened here with code 0x0
    │
    =     at ./sources/aggregator/optional_aggregator.move:134: destroy_optional_aggregator
    =         optional_aggregator =
    =           optional_aggregator.OptionalAggregator{
    =             aggregator =
    =               option.Option{
    =                 vec =
    =                   vector{
    =                     aggregator.Aggregator{
    =                       handle = 0x18be,
    =                       key = 0x4823,
    =                       limit = 41u128}}},
    =             integer = option.Option{vec = vector{}}}
    =     at ./sources/aggregator/optional_aggregator.move:135: destroy_optional_aggregator
    =         integer = option.Option{vec = vector{}}
    =         aggregator =
    =           option.Option{
    =             vec =
    =               vector{
    =                 aggregator.Aggregator{
    =                   handle = 0x18be,
    =                   key = 0x4823,
    =                   limit = 41u128}}}
    =     at ./sources/aggregator/optional_aggregator.move:136: destroy_optional_aggregator
    =     at ./sources/aggregator/aggregator.move:80: limit
    =         aggregator =
    =           aggregator.Aggregator{handle = 0x18be, key = 0x4823, limit = 41u128}
    =     at ./sources/aggregator/aggregator.move:81: limit
    =         result = 41u128
    =     at ./sources/aggregator/aggregator.move:82: limit
    =         limit = 41u128
    =     at ./sources/aggregator/optional_aggregator.move:137: destroy_optional_aggregator
    =         ABORTED

error: abort not covered by any of the `aborts_if` clauses
    ┌─ ./sources/aggregator/optional_aggregator.spec.move:42:5
    │
 42 │ ╭     spec switch_and_zero_out(optional_aggregator: &mut OptionalAggregator) {
 43 │ │         let vec_ref = optional_aggregator.integer.vec;
 44 │ │         aborts_if is_parallelizable(optional_aggregator) && len(vec_ref) != 0;
 45 │ │         aborts_if !is_parallelizable(optional_aggregator) && len(vec_ref) == 0;
 46 │ │         aborts_if !is_parallelizable(optional_aggregator) && !exists<aggregator_factory::AggregatorFactory>(@aptos_framework);
 47 │ │     }
    │ ╰─────^
    │
    ┌─ ./sources/aggregator/optional_aggregator.move:105:9
    │
105 │         aggregator::destroy(aggregator);
    │         ------------------------------- abort happened here with code 0xA
    │
    =     at ./sources/aggregator/optional_aggregator.move:90: destroy
    =     at ./sources/aggregator/optional_aggregator.spec.move:43
    =     at ./sources/aggregator/optional_aggregator.move:90: destroy
    =         optional_aggregator =
    =           &optional_aggregator.OptionalAggregator{
    =             aggregator =
    =               option.Option{
    =                 vec =
    =                   vector{
    =                     aggregator.Aggregator{
    =                       handle = 0x4ae1,
    =                       key = 0x6784,
    =                       limit = 6334u128}}},
    =             integer = option.Option{vec = vector{}}}
    =     at ./sources/aggregator/optional_aggregator.move:91: switch_and_zero_out
    =     at ./sources/aggregator/optional_aggregator.move:185: is_parallelizable
    =         optional_aggregator =
    =           optional_aggregator.OptionalAggregator{
    =             aggregator =
    =               option.Option{
    =                 vec =
    =                   vector{
    =                     aggregator.Aggregator{
    =                       handle = 0x4ae1,
    =                       key = 0x6784,
    =                       limit = 6334u128}}},
    =             integer = option.Option{vec = vector{}}}
    =     at ./sources/aggregator/optional_aggregator.move:186: is_parallelizable
    =         result = true
    =     at ./sources/aggregator/optional_aggregator.move:187: is_parallelizable
    =     at ./sources/aggregator/optional_aggregator.move:92: switch_and_zero_out
    =     at ./sources/aggregator/optional_aggregator.move:100: switch_to_integer_and_zero_out
    =         optional_aggregator =
    =           &optional_aggregator.OptionalAggregator{
    =             aggregator =
    =               option.Option{
    =                 vec =
    =                   vector{
    =                     aggregator.Aggregator{
    =                       handle = 0x4ae1,
    =                       key = 0x6784,
    =                       limit = 6334u128}}},
    =             integer = option.Option{vec = vector{}}}
    =     at ./sources/aggregator/optional_aggregator.move:103: destroy_optional_integer
    =     at <internal>:1
    =     at ./sources/aggregator/optional_aggregator.move:103: destroy_optional_integer
    =         optional_aggregator =
    =           &optional_aggregator.OptionalAggregator{
    =             aggregator = option.Option{vec = vector{}},
    =             integer = option.Option{vec = vector{}}}
    =         aggregator =
    =           aggregator.Aggregator{
    =             handle = 0x4ae1,
    =             key = 0x6784,
    =             limit = 6334u128}
    =     at ./sources/aggregator/optional_aggregator.move:104: destroy_optional_integer
    =     at ./sources/aggregator/aggregator.move:80: limit
    =         aggregator =
    =           aggregator.Aggregator{
    =             handle = 0x4ae1,
    =             key = 0x6784,
    =             limit = 6334u128}
    =     at ./sources/aggregator/aggregator.move:81: limit
    =         result = 6334u128
    =     at ./sources/aggregator/aggregator.move:82: limit
    =         limit = 6334u128
    =     at ./sources/aggregator/optional_aggregator.move:105: switch_to_integer_and_zero_out
    =     at <internal>:1
    =     at ./sources/aggregator/optional_aggregator.move:105: switch_to_integer_and_zero_out
    =         ABORTED

error: abort not covered by any of the `aborts_if` clauses
    ┌─ ./sources/aggregator/optional_aggregator.spec.move:50:5
    │
 50 │ ╭     spec switch_to_integer_and_zero_out(
 51 │ │         optional_aggregator: &mut OptionalAggregator
 52 │ │     ): u128 {
 53 │ │         aborts_if len(optional_aggregator.aggregator.vec) == 0;
 54 │ │         aborts_if len(optional_aggregator.integer.vec) != 0;
 55 │ │     }
    │ ╰─────^
    │
    ┌─ ./sources/aggregator/optional_aggregator.move:105:9
    │
105 │         aggregator::destroy(aggregator);
    │         ------------------------------- abort happened here with code 0x0
    │
    =     at ./sources/aggregator/optional_aggregator.move:100: switch_to_integer_and_zero_out
    =         optional_aggregator =
    =           &optional_aggregator.OptionalAggregator{
    =             aggregator =
    =               option.Option{
    =                 vec =
    =                   vector{
    =                     aggregator.Aggregator{
    =                       handle = 0x6784,
    =                       key = 0x18be,
    =                       limit = 18467u128}}},
    =             integer = option.Option{vec = vector{}}}
    =     at ./sources/aggregator/optional_aggregator.move:103: destroy_optional_integer
    =     at <internal>:1
    =     at ./sources/aggregator/optional_aggregator.move:103: destroy_optional_integer
    =         optional_aggregator =
    =           &optional_aggregator.OptionalAggregator{
    =             aggregator = option.Option{vec = vector{}},
    =             integer = option.Option{vec = vector{}}}
    =         aggregator =
    =           aggregator.Aggregator{
    =             handle = 0x6784,
    =             key = 0x18be,
    =             limit = 18467u128}
    =     at ./sources/aggregator/optional_aggregator.move:104: destroy_optional_integer
    =     at ./sources/aggregator/aggregator.move:80: limit
    =         aggregator =
    =           aggregator.Aggregator{
    =             handle = 0x6784,
    =             key = 0x18be,
    =             limit = 18467u128}
    =     at ./sources/aggregator/aggregator.move:81: limit
    =         result = 18467u128
    =     at ./sources/aggregator/aggregator.move:82: limit
    =         limit = 18467u128
    =     at ./sources/aggregator/optional_aggregator.move:105: switch_to_integer_and_zero_out
    =     at <internal>:1
    =     at ./sources/aggregator/optional_aggregator.move:105: switch_to_integer_and_zero_out
    =         ABORTED

Error: exiting with verification errors
```

Reply:
Something is not right here. Why does `destroy` aborts with different error codes? cc @0xOutOfGas @junkil-park 

Also, I am curious how native functions are verified? In `destroy`, we guarantee that `AggregatorFactory` contains the key of the aggregator pointing to some value. Not 100% suse how to write this for prover but something like

```
spec destroy(aggregator: Aggregator) {
   aborts_if !exists<AggregatorFactory>(@aptos_framework);
   aborts_if !table::contains(borrow_global<AggregatorFactory>(@aptos_framework).phantom_table, aggregator.key);
}
```

Reply:
Different error codes, may be caused by several function calls, the `destroy` is called by  `optional_aggregator::switch_to_integer_and_zero_out` and `optional_aggregator::destroy_optional_aggregator`, we have specs for these two functions.

Reply:
@georgemitenkov, that indeed looks a right model for `destroy`! However, it's not possible to specify it because `AggregatorFactory` cannot be referred to in the `aggregator` module due to the cyclic dependency. In this case, modeling the native functions of `aggregator` should be done in a lower level (i.e., boogie level). It may need to be done in a separate PR.

@0xOutOfGas , @qpb8023 , in the meantime, could you model it using a uninterpreted spec function like:
```
    spec fun spec_aggreator_exists(aggregator: Aggregator): bool;

    spec destroy(aggregator: Aggregator) {
        pragma opaque;
        aborts_if !spec_aggreator_exists(aggreator);
    }
```
And, add `aborts_if !spec_aggreator_exists(aggreator);` to the specs of `add`, `sub` and `read` as well?

Reply:
> 'not implemented: Not yet supported constant vector type: Vector(Primitive(U64))'

This seems a separate issue. Do you still have the same problem when using the latest move-cli?

Reply:
@rahxephon89, we support the constant vector type like `Vector(Primitive(U64))`, right?

Reply:
Just to unblock this PR from the constant vector issue, can you use this spec and proceed?
```
spec destroy(aggregator: Aggregator) {
        pragma opaque;
        // TODO: this aborts_if condition is a temporary mockup, which needs to be refined.
        aborts_if false;
    }
```